Tree widget

namreg opened the pull request:
Hello, there.
I've created a tree widget for my project. I believe this can be useful for anyone else :)
Let me know, please, what you think about that.

Thanks for the termui project!

namreg commented:
/cc @cjbassi

srlehn commented:
screenshot of the example:
![Auswahl_2019 04 19_12:07:52](https://user-images.githubusercontent.com/48837958/56419851-003e1680-629c-11e9-8e75-e190ca99d4f3.png)

cjbassi commented:
Thanks for the pr! I'd like to merge this but I'll need some time to look over it and possibly suggest any changes.

cjbassi commented on theme.go:
```diff
@@ -66,6 +67,12 @@ type ListTheme struct {
+type TreeTheme struct {
+	Text     Style
+	Collapse rune
+	Expand   rune
```
Rename `Collapse` to `Collapsed` and `Expand` to `Expanded`.

cjbassi commented on widgets/tree.go:
```diff
@@ -0,0 +1,242 @@
+}
+
+func (self *Tree) prepareNodes() {
+	self.rows = self.rows[:0]
```
Is this emptying the slice? A comment explaining this would be good.

cjbassi commented on widgets/tree.go:
```diff
@@ -0,0 +1,242 @@
+	SelectedRow      int
+	topRow           int
+
+	//rows is flatten nodes for rendering.
```
Fix space before comment text.

cjbassi commented on widgets/tree.go:
```diff
@@ -0,0 +1,242 @@
+	Expanded bool
+	Nodes    []*TreeNode
+
+	//level stores the node level in the tree.
```
Fix space before comment text.

cjbassi commented on widgets/tree.go:
```diff
@@ -0,0 +1,242 @@
+	SelectedRowStyle Style
+	WrapText         bool
+	SelectedRow      int
+	topRow           int
```
Move this field down a line next to the other private fields.

namreg commented:
@cjbassi 
> 1. How should the cursor be positioned after collapsing a node? Currently it seems like SelectedRow stays the same, but sometimes that leaves no currently visible node selected. Although that might be fine. Just wanted to get some thoughts.

I did it intentionally. IMHO, cursor position relate to the line in block area, not to the content is rendered to this area. May be I wrong.

> 2. Expand and the other functions that operate on self.rows would potentially operate on an incorrect node if a node is added and Expand is called before the tree is rerendered, since self.rows is only updated in Draw. What if we made self.Nodes private and added a SetNodes method that calls prepareNodes? That would also eliminate calling prepareNodes every draw too.

It like the idea to add public method `SetNodes` and make property `rows` private. But we can't rid of the `prepareNodes` calling before render after that because `self.rows` contains only visible rows (in our case `expanded`). Hence, I have moved calling of the `prepareNodes` to every place where `self.rows` is changing. Please, let me know what you think about that.

namreg commented:
gentle ping @cjbassi

cjbassi commented:
Apologies for the delay, but I haven't forgotten about this! Currently finishing up the school quarter and I should be able to get to this soon.

namreg commented:
No problem) Good luck in the school!

On Sat, 8 Jun 2019 at 01:29, Caleb Bassi <notifications@github.com> wrote:

> Apologies for the delay, but I haven't forgotten about this! Currently
> finishing up the school quarter and I should be able to get to this soon.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gizak/termui/pull/237?email_source=notifications&email_token=AAC7VQCTB4GANFAKG4F5NKTPZLOLHA5CNFSM4HG547Q2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXHEVUA#issuecomment-500058832>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAC7VQCKD4BKOMW3UL4RWRDPZLOLHANCNFSM4HG547QQ>
> .
>

cjbassi commented:
Sorry for the delay again! Been out of commission a bit due to an injury. I've looked it over again and it LGTM. Changing the widget so that nodes are set with `SetNodes` looks good. I had to rebase to fix some conflicts but I just merged this and now I'm going to release 3.1. Thanks!

cjbassi commented:
I accidentally merged before force pushing the PR which is why it says the PR was closed just btw.
